src/l2_tracker/events.h

```diff
@@ -16,13 +16,21 @@ using namespace std::literals;
 namespace eth::event {
 
 struct L2StateChange {
-    uint64_t l2_height = 0;
+    uint64_t chain_id;
+    uint64_t l2_height;
 
     std::strong_ordering operator<=>(const L2StateChange&) const = default;
 
   protected:
-    L2StateChange() = default;
-    L2StateChange(uint64_t l2_height) : l2_height{l2_height} {}
+    L2StateChange(uint64_t chain_id, uint64_t l2_height) :
+            chain_id{chain_id}, l2_height{l2_height} {}
+
+    template <class Archive>
+    void serialize_base(Archive& ar, uint8_t version) {
```

Comment:
Version needs to passed in as a reference `uint8_t &version` so that calling code can branch on the version if it changes to conditionally deserialise the binary stream.

```
void serialize_base(Archive& ar, uint8_t version) {
```

Though I don't see a situation where the base `L2StateChange` would change between different ethereum events that we can just inline the version.

```cpp
        uint8_t version = 0;
        field_varint(ar, "version", version);
        field_varint(ar, "chain_id", chain_id);
        field_varint(ar, "l2_height", l2_height);
```

I'd also just include a version in the derived class as well to track changes in those structs independently from the base class. It'll only cost an additional byte to store the version for each TX extra.

Reply:
In retrospect trying to split the serialization into the base class wasn't worth the complexity, so I've just moved the two fields into each derived class and dropped the base class serialization code entirely.